pkg/kafka/partitionring/consumer/client.go

```diff
@@ -0,0 +1,105 @@
+package consumer
+
+import (
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/ring"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+
+	"github.com/grafana/loki/v3/pkg/kafka"
+	"github.com/grafana/loki/v3/pkg/kafka/client"
+)
+
+type Client struct {
+	*kgo.Client
+	partitionRing ring.PartitionRingReader
+	logger        log.Logger
+	stopCh        chan struct{}
+	wg            sync.WaitGroup
+}
+
+// NewGroupClient creates a new Kafka consumer group client that participates in cooperative group consumption.
+// It joins the specified consumer group and consumes messages from the configured Kafka topic.
+//
+// The client uses a cooperative-active-sticky balancing strategy which ensures active partitions are evenly
+// distributed across group members while maintaining sticky assignments for optimal processing. Inactive partitions
+// are still assigned and monitored but not processed, allowing quick activation when needed. Partition handoffs
+// happen cooperatively to avoid stop-the-world rebalances.
+//
+// The client runs a background goroutine that monitors the partition ring for changes. When the set of active
+// partitions changes (e.g. due to scaling or failures), it triggers a rebalance to ensure partitions are
+// properly redistributed across the consumer group members. This maintains optimal processing as the active
+// partition set evolves.
+func NewGroupClient(kafkaCfg kafka.Config, partitionRing ring.PartitionRingReader, groupName string, metrics *kprom.Metrics, logger log.Logger, opts ...kgo.Opt) (*Client, error) {
+	defaultOpts := []kgo.Opt{
+		kgo.ConsumerGroup(groupName),
+		kgo.ConsumeTopics(kafkaCfg.Topic),
+		kgo.Balancers(NewCooperativeActiveStickyBalancer(partitionRing)),
```

Comment:
nit. Shouldn't make this not overridable by `opts`?

Reply:
I rather have sane default and allow them to be overridden is needed.